main.c

```diff
@@ -0,0 +1,294 @@
+    }
+  if ((private_key = PEM_read_bio_RSAPrivateKey (bio, 0, 0, 0)) == NULL)
+    {
+      napi_throw_error (env, NULL, "Failed to read key from bio");
+      BIO_free (bio);
+      free (buf);
+      return NULL;
+    }
+
+  if (private_key->n == NULL || private_key->e == NULL ||
+      private_key->d == NULL || private_key->p == NULL ||
+      private_key->q == NULL || private_key->dmp1 == NULL ||
+      private_key->dmq1 == NULL || private_key->iqmp == NULL)
+    {
+      napi_throw_error (env, NULL, "One or more required values in the "
+                        "private key is null");
+      RSA_free (private_key);
+      BIO_free (bio);
+      free (buf);
+      return NULL;
+    }
+
+  n_hex = BN_bn2hex (private_key->n);
+  e_hex = BN_bn2hex (private_key->e);
+  d_hex = BN_bn2hex (private_key->d);
+  p_hex = BN_bn2hex (private_key->p);
+  q_hex = BN_bn2hex (private_key->q);
+  dmp1_hex = BN_bn2hex (private_key->dmp1);
+  dmq1_hex = BN_bn2hex (private_key->dmq1);
+  iqmp_hex = BN_bn2hex (private_key->iqmp);
+
+  BIO_free (bio);
+  RSA_free (private_key);
+  free (buf);
+
+  status |= napi_create_string_utf8 (env, n_hex, strlen (n_hex), &n_val);
+  status |= napi_create_string_utf8 (env, e_hex, strlen (e_hex), &e_val);
+  status |= napi_create_string_utf8 (env, d_hex, strlen (d_hex), &d_val);
+  status |= napi_create_string_utf8 (env, p_hex, strlen (p_hex), &p_val);
+  status |= napi_create_string_utf8 (env, q_hex, strlen (q_hex), &q_val);
+  status |= napi_create_string_utf8 (env, dmp1_hex, strlen (dmp1_hex), &dmp1_val);
+  status |= napi_create_string_utf8 (env, dmq1_hex, strlen (dmq1_hex), &dmq1_val);
+  status |= napi_create_string_utf8 (env, iqmp_hex, strlen (iqmp_hex), &iqmp_val);
+  status |= napi_create_object (env, &obj);
+
+  free (n_hex);
+  free (e_hex);
+  free (d_hex);
+  free (p_hex);
+  free (q_hex);
+  free (dmp1_hex);
+  free (dmq1_hex);
+  free (iqmp_hex);
+
+  if (status != napi_ok)
+    {
+      napi_throw_error (env, NULL, "Failed to create return object");
+      return NULL;
+    }
+
+  status |= napi_set_named_property (env, obj, "n", n_val);
+  status |= napi_set_named_property (env, obj, "e", e_val);
+  status |= napi_set_named_property (env, obj, "d", d_val);
+  status |= napi_set_named_property (env, obj, "p", p_val);
+  status |= napi_set_named_property (env, obj, "q", q_val);
+  status |= napi_set_named_property (env, obj, "dmp1", dmp1_val);
+  status |= napi_set_named_property (env, obj, "dmq1", dmq1_val);
+  status |= napi_set_named_property (env, obj, "iqmp", iqmp_val);
+
+  if (status != napi_ok)
+    {
+      napi_throw_error (env, NULL, "Failed to assign properties to object");
+      return NULL;
+    }
+
+  return obj;
+}
+
+static napi_value
+init (napi_env env, napi_value exports) {
+  napi_value rsa_fn, x509_fn;
+
+  if (napi_create_function (env, NULL, 0, rsa_priv_key,
+                            NULL, &rsa_fn) != napi_ok)
+    napi_throw_error (env, NULL, "Unable to wrap native rsa function");
```

Comment:
These should have return statements all of them. I don't think it stop when it hits an error